File: Frontend/src/app/shared/service/usuario.service.ts
import { Injectable } from "@angular/core";
import { Usuario } from '../model/usuario';
import { HttpClient } from '@angular/common/http';
import { Observable } from "rxjs";
import { UsuarioFiltro } from "../model/filtros/usuario.filtro";
import { Bloco } from "../model/bloco";
import { Apartamento } from "../model/apartamento";

@Injectable({
  providedIn: 'root'
})
export class UsuarioService{

  private readonly API = 'http://localhost:8080/SmartShop/rest/usuario';

  constructor(
    private httpClient: HttpClient
  ) { }

  consultarTodos(): Observable<Array<Usuario>> {
    return this.httpClient.get<Array<Usuario>>(this.API + '/todos');
  }

  salvar(usuario: Usuario): Observable<Usuario> {
    return this.httpClient.post<Usuario>(this.API, usuario);
  }

  atualizar(usuario: Usuario): Observable<boolean> {
    return this.httpClient.put<boolean>(this.API + "/alterar", usuario);
  }

  buscar(id: number): Observable<Usuario> {
    return this.httpClient.get<Usuario>(`${this.API}/buscar/${id}`);
  }

  buscarCpf(cpf: string): Observable<Usuario> {
    return this.httpClient.get<Usuario>(this.API + "/buscarcpf/" + cpf);
  }

  excluir(id: number): Observable<boolean> {
    return this.httpClient.delete<boolean>(this.API + "/excluir/" + id);
  }

  consultarComFiltro(filtro: UsuarioFiltro): Observable<Array<Usuario>> {
    return this.httpClient.post<Array<Usuario>>(`${this.API}/filtrar`, filtro);
  }

  contarRegistro(filtro: UsuarioFiltro): Observable<number> {
    return this.httpClient.post<number>(this.API + "/contarRegistros", filtro);
  }

  contarPaginas(filtro: UsuarioFiltro): Observable<number> {
    return this.httpClient.post<number>(this.API + "/totalPaginas", filtro);
  }

  consultarBlocoPorNome(nome: string): Observable<Bloco> {
    return this.httpClient.get<Bloco>(`${this.API}/bloco/nome/${nome}`);
  }

  consultarApartamentoPorNumero(numero: number): Observable<Apartamento> {
    return this.httpClient.get<Apartamento>(`${this.API}/apartamento/numero/${numero}`);
  }
}

File: back-end/src/main/java/service/UsuarioService.java
package service;

import java.util.ArrayList;
import java.util.List;

import exception.SmartshopException;
import model.entity.Apartamento;
import model.entity.Bloco;
import model.entity.Usuario;
import model.entity.filtros.UsuarioFiltro;
import model.repository.UsuarioRepository;

public class UsuarioService {

    private UsuarioRepository repository = new UsuarioRepository();

    public Usuario salvar(Usuario novoUsuario) throws SmartshopException {
        if (repository.verificarUsuarioExiste(novoUsuario)) {
            throw new SmartshopException("Usuário com o email " + novoUsuario.getEmail() + " já existe.");
        } else {
            return repository.salvar(novoUsuario);
        }
    }

    public boolean atualizar(Usuario usuarioEditado) {
        return repository.alterar(usuarioEditado);
    }

    public boolean excluir(int id) throws SmartshopException {
        return repository.excluir(id);
    }

    public Usuario consultarPorId(int id) {
        return repository.consultarPorId(id);
    }
    
    public Usuario consultarPorCpf(String cpf) {
    	return repository.consultarPorCpf(cpf);
    }

    public List<Usuario> consultarTodos() {
        return repository.consultarTodos();
    }

    public ArrayList<Usuario> buscarComFiltro(UsuarioFiltro filtro) {
        return repository.buscarComFiltro(filtro);
    }

    public int contarRegistros(UsuarioFiltro filtro) {
        return repository.contarRegistros(filtro);
    }


    public Bloco consultarBlocoPorNome(String nome) {
        return repository.consultarBlocoPorNome(nome);
    }

    public Apartamento consultarApartamentoPorNumero(int numero) {
        return repository.consultarApartamentoPorNumero(numero);
    }
}
